### ISSUE TITLE
Why is it wrong to use the cursor within brace in `Parse::parse` fn?

### ISSUE THREAD

**Issue by zjp-CN** [dtolnay/syn#1120]
The following code generates a token stream well, but still doesn't work:

```rust
    pub struct Seq {
        ident:       syn::Ident,
        brace_token: syn::token::Brace,
        pub tokens:  TokenStream2,
    }


    impl syn::parse::Parse for Seq {
        fn parse(input: syn::parse::ParseStream) -> syn::Result<Self> {
            let ident = input.parse()?;
            let content;
            let brace_token = syn::braced!(content in input);

            // *** Here is the difference ***
            let cursor = content.cursor();
            let tokens = tokens(cursor);
            dbg!(&tokens);
            // ******************************

            Ok(Seq { ident,
                     brace_token,
                     tokens })
        }
    }
```

The result and whole MWE is [here](https://github.com/zjp-CN/hello-syn).

So is the following style is the recommanded way to parse a custom syntax?

```rust
    pub struct Seq {
        ident:       syn::Ident,
        brace_token: syn::token::Brace,
        tokens:      TokenStream2,
    }


    impl syn::parse::Parse for Seq {
        fn parse(input: syn::parse::ParseStream) -> syn::Result<Self> {
            let ident = input.parse()?;
            let content;
            let brace_token = syn::braced!(content in input);

            // *** Here is the difference ***
            let tokens = content.parse()?;
            // ******************************

            Ok(Seq { ident,
                     brace_token,
                     tokens })
        }
    }


    impl Seq {
        pub fn tokens(self) -> TokenStream2 {
            let buf = syn::buffer::TokenBuffer::new2(self.tokens);
            tokens(buf.begin())
        }
    }
```

**Comment by dtolnay** [comment]
This is what the documentation of `ParseStream::cursor` means where it says:

> *Cursors are immutable so no operations you perform against the cursor will affect the state of this parse stream.*

Whatever you do with a `Cursor` isn't gonna result in consuming any tokens from the input. Since you are not parsing any tokens from the input stream, you'll see "unexpected token" error on all the tokens that you did not parse.

**Comment by gagbo** [comment]
Hello,

Sorry for piggy-backing on this (next paragraph is optional I guess), I'm currently doing your `proc-macro-workshop` (thanks for that), and struggling to understand what the idiomatic solution to this problem would be. 

The linked commit above used a small custom `ParseBuffer::step` iteration to be sure to get all the braced content in the stream, but the documentation of syn specifies that it is [unlikely to have to use this function](https://docs.rs/syn/1.0.99/syn/parse/struct.ParseBuffer.html#method.step) (I tried to just use a single `content.parse()` call as the sample code in the OP, but I only get the first token of the braced content). Is there a method or a trait I'm missing either from proc_macro2 or syn to extract a full delimited content as a TokenStream?

tl;dr: Is `ParseBuffer::step` the only way to extract a TokenStream from `delimited!` macros ? Also, do you know a location where some idiomatic implementations of the workshop can be found ? I'm really longing for some experimented solution to compare my hacky stuff to once I'm done exploring the API.

**Comment by zjp-CN** [comment]
> Is `ParseBuffer::step` the only way to extract a TokenStream from delimited! macros ?

No. A `.parse()` method can do the same thing here.

> Is there a method or a trait I'm missing either from proc_macro2 or syn to extract a full delimited content as a TokenStream?

That's the functionality of [`Parse`](https://docs.rs/syn/latest/syn/parse/trait.Parse.html) trait in `Syn`.

---

The problem I ran into is because:
* Although `let cursor = content.cursor();` yields a starting cursor from `ParseStream` and is parsed to `TokenStream`, which is a way to parse something
* the cursor is consumed (wether till the end or not), but the `ParseStream`'s internal cursor is unchanged, which leads to some tokens not parsed and causes the error.


So `ParseBuffer::step` and `.parse()` is correct because they can push the cursor to the end if successful.
And 
* the former exposes its inner cursor ([in some way](https://docs.rs/syn/1.0.99/src/syn/parse.rs.html#1006-1034)), which means you can directly check the tokens
* but the latter needs a predefined type (to be returned), and you often make use of most the Syn's predefined types, which are implemented with `Parse` trait, in your custom-defined type.
